lib/PhasarLLVM/ControlFlow/Resolver/OTFResolver.cpp

```diff
@@ -144,18 +140,74 @@ OTFResolver::resolveFunctionPointer(llvm::ImmutableCallSite CS) {
         if (P->getType()->isPointerTy() &&
             P->getType()->getPointerElementType()->isFunctionTy()) {
           if (const auto *F = llvm::dyn_cast<llvm::Function>(P)) {
-            if (matchesSignature(F, FTy)) {
+            if (matchesSignature(F, FTy, false)) {
               Callees.insert(F);
             }
           }
         }
+        std::vector<const llvm::GlobalVariable *> GlobalVariableWL;
+        std::stack<const llvm::ConstantAggregate *> ConstantAggregateWL;
+        if (auto *CE = llvm::dyn_cast<llvm::ConstantExpr>(P)) {
+          // Unfortunately this allocates
+          auto *AsI = CE->getAsInstruction();
+          for (auto &Op : AsI->operands()) {
+            if (auto *GVOp = llvm::dyn_cast<llvm::GlobalVariable>(Op)) {
+              GlobalVariableWL.push_back(GVOp);
+            }
+          }
+          AsI->deleteValue();
+        }
+        if (auto *GVP = llvm::dyn_cast<llvm::GlobalVariable>(P)) {
+          GlobalVariableWL.push_back(GVP);
+        }
+        for (auto *GV : GlobalVariableWL) {
+          if (!GV->hasInitializer()) {
+            continue;
+          }
+          auto InitConst = GV->getInitializer();
+          if (auto *InitConstAggregate =
+                  llvm::dyn_cast<llvm::ConstantAggregate>(InitConst)) {
+            ConstantAggregateWL.push(InitConstAggregate);
+          }
+        }
+        while (!ConstantAggregateWL.empty()) {
+          auto ConstAggregateItem = ConstantAggregateWL.top();
+          ConstantAggregateWL.pop();
+          for (const auto &Op : ConstAggregateItem->operands()) {
+            if (auto *CE = llvm::dyn_cast<llvm::ConstantExpr>(Op)) {
+              auto *AsI = CE->getAsInstruction();
+              if (AsI->getType()->getPointerElementType() == FTy) {
+                if (auto *BC = llvm::dyn_cast<llvm::BitCastInst>(AsI)) {
+                  if (auto *F =
+                          llvm::dyn_cast<llvm::Function>(BC->getOperand(0))) {
+                    Callees.insert(F);
+                  }
+                }
+              }
+              AsI->deleteValue();
+            }
+            if (auto *F = llvm::dyn_cast<llvm::Function>(Op)) {
+              if (matchesSignature(F, FTy, false)) {
+                Callees.insert(F);
+              }
+            }
+            if (auto *CA = llvm::dyn_cast<llvm::ConstantAggregate>(Op)) {
+              ConstantAggregateWL.push(CA);
+            }
+            if (auto *GV = llvm::dyn_cast<llvm::GlobalVariable>(Op)) {
+              if (!GV->hasInitializer()) {
+                continue;
+              }
+              if (auto *GVCA = llvm::dyn_cast<llvm::ConstantAggregate>(
+                      GV->getInitializer())) {
+                ConstantAggregateWL.push(GVCA);
```

Comment:
I believe there's a possibility for an infinite loop here with the pushing things onto the ConstantAggregateWL.

Reply:
Could you provide your sample program please? This part here is supposed to recurse into nested global objects, which should be finite.

Reply:
Oops, I used the wrong github account on my initial review!
I'll have to create sample, the program I was analyzing is not something I have the ability to share.  But I can say what I saw from GDB was the variable "GVCA" at line 203 was the same as "ConstAggregateItem" from line 174, thus the logic would not exit the while loop since ConstantAggregateWL never becomes empty.  I can look into this more in a few days and give you more specifics and/or some sample code.

Reply:
It would suffice to isolate the data structure where this occurs. That would show how a ConstantAggregate can have itself as an initializer of its own operand, which seems very weird to me.

Reply:
Good news, I was able to track down where this was happening in the application this morning and it is in an open source library.

Super short version -- I believe this is the struct in question:
```
struct metal_list {
	struct metal_list *next, *prev;
};
```

In https://github.com/OpenAMP/libmetal/blob/master/lib/irq.c  the variable irq_cntrs which is declared as part of the macro on line 14.  That macro is in https://github.com/OpenAMP/libmetal/blob/master/lib/list.h line 38/39, with the related code from lines 25-33.

Here's some gdb output that may be helpful:

```
(gdb) bt
#0  psr::OTFResolver::resolveFunctionPointer (this=0x5555565289c0, CS=...) at ../lib/PhasarLLVM/ControlFlow/Resolver/OTFResolver.cpp:206
#1  0x00007ffff7e59e2a in psr::LLVMBasedICFG::constructDynamicCall (this=0x7fffffffdf28, I=0x555556554cf0, Resolver=...) at ../lib/PhasarLLVM/ControlFlow/LLVMBasedICFG.cpp:285
#2  0x00007ffff7e5b093 in psr::LLVMBasedICFG::LLVMBasedICFG (this=0x7fffffffdf28, IRDB=..., CGType=psr::CallGraphAnalysisType::OTF, EntryPoints=std::set with 1 element = {...}, TH=0x7fffffffdd90, 
    PT=0x55555655b920, SF=psr::SoundnessFlag::SOUNDY) at ../lib/PhasarLLVM/ControlFlow/LLVMBasedICFG.cpp:134
#3  0x000055555558ac08 in ArtusPhasar::CallgraphContainer::CallgraphContainer(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, psr::IRDBOptions, psr::CallGraphAnalysisType, std::set<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&) ()
#4  0x000055555556c5d8 in main ()
Breakpoint 1, psr::OTFResolver::resolveFunctionPointer (this=0x5555565289c0, CS=...) at ../lib/PhasarLLVM/ControlFlow/Resolver/OTFResolver.cpp:206
206                     ConstantAggregateWL.push(GVCA);
(gdb) print GVCA
$1 = (llvm::ConstantAggregate *) 0x5555556ad500
(gdb) print ConstAggregateItem
$2 = (const llvm::ConstantAggregate *) 0x5555556ad500
(gdb) print *GV
$5 = {<llvm::GlobalObject> = {<llvm::GlobalValue> = {<llvm::Constant> = {<llvm::User> = {<llvm::Value> = {VTy = 0x5555555ff940, UseList = 0x55555657d870, SubclassID = 3 '\003', HasValueHandle = 0 '\000', 
            SubclassOptionalData = 0 '\000', SubclassData = 0, NumUserOperands = 1, IsUsedByMD = 0, HasName = 1, HasHungOffUses = 0, HasDescriptor = 0, static MaxAlignmentExponent = 29, 
            static MaximumAlignment = 536870912}, <No data fields>}, <No data fields>}, ValueType = 0x5555555fd190, static GlobalValueSubClassDataBits = 16, Linkage = 7, Visibility = 0, UnnamedAddrVal = 0, 
      DllStorageClass = 0, ThreadLocal = 0, HasLLVMReservedName = 0, IsDSOLocal = 1, HasPartition = 0, SubClassData = 36, IntID = 0, Parent = 0x5555555fa180}, ObjComdat = 0x0, 
    static GlobalObjectSubClassDataBits = 9, static AlignmentBits = 5, static AlignmentMask = 31, 
    static GlobalObjectMask = 127}, <llvm::ilist_node<llvm::GlobalVariable>> = {<llvm::ilist_node_impl<llvm::ilist_detail::node_options<llvm::GlobalVariable, false, false, void> >> = {<llvm::ilist_node_base<false>> = {Prev = 0x5555556ad3a0, Next = 0x5555555fa188}, <No data fields>}, <No data fields>}, Attrs = {SetNode = 0x0}, isConstantGlobal = false, isExternallyInitializedConstant = false}
(gdb) print GV->getName()
$6 = {static npos = 18446744073709551615, Data = 0x5555556ad490 "irq_cntrs", Length = 9}
(gdb) c
Continuing.

Breakpoint 1, psr::OTFResolver::resolveFunctionPointer (this=0x5555565289c0, CS=...) at ../lib/PhasarLLVM/ControlFlow/Resolver/OTFResolver.cpp:206
206                     ConstantAggregateWL.push(GVCA);
(gdb) print GVCA
$7 = (llvm::ConstantAggregate *) 0x5555556ad500
(gdb) print ConstantAggregateWL
$9 = std::stack wrapping: std::deque with 2593485 elements = {0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 
  0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 
  0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 
  0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 
  0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 
  0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 
  0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 
  0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 
  0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 
  0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 
  0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 
  0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 
  0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 
  0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 
  0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 
  0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500, 0x5555556ad500...}
```

Reply:
Also, I found that the test app that comes with libmetal has the same issue as above, and can be used to reproduce the problem with irq_cntrs.  libmetal can be built with wllvm, and the extract the .ll file from test-metal-static for use.

Reply:
Thank you, David, I overlooked this case. I'll look into this again.